sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -413,6 +418,10 @@ object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
         case LessThan(l, r) => GreaterThanOrEqual(l, r)
         // not(l <= r)  =>  l > r
         case LessThanOrEqual(l, r) => GreaterThan(l, r)
+        // not(l || r) => not(l) && not(r)
+        case Or(l, r) => And(Not(l), Not(r))
+        // not(l && r) => not(l) or not(r)
+        case And(l, r) => Or(Not(l), Not(r))
```

Comment:
At first I thought this optimization should be only used in datasource as it may cause performance regression for normal cases(see these [comments](https://github.com/apache/spark/pull/8200#discussion-diff-37154582)). However after an offline discussion with @marmbrus , this optimization is pretty standard and generally good, so let's keep it in `Optimizer` :)